fendermint/vm/message/src/ipc.rs

```diff
@@ -0,0 +1,169 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use cid::Cid;
+use fvm_shared::{address::Address, clock::ChainEpoch, crypto::signature::Signature};
+use ipc_sdk::subnet_id::SubnetID;
+use serde::{Deserialize, Serialize};
+
+/// Messages involved in InterPlanetary Consensus.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[allow(clippy::large_enum_variant)]
+pub enum IpcMessage {
+    /// A bottom-up checkpoint coming from a child subnet "for resolution", relayed by a user of the parent subnet for a reward.
+    ///
+    /// The reward can be given immediately upon the validation of the quorum certificate in the checkpoint,
+    /// or later during execution, once data availability has been confirmed.
+    BottomUpResolve(SignedRelayedMessage<CertifiedMessage<BottomUpCheckpoint>>),
+
+    /// A bottom-up checkpoint proposed "for execution" by the parent subnet validators, provided that the majority of them
+    /// have the data available to them, already resolved.
+    ///
+    /// To prove that the data is available, we can either use the ABCI++ "process proposal" mechanism,
+    /// or we can gossip votes using the _IPLD Resolver_ and attach them as a quorum certificate.
+    BottomUpExec(CertifiedMessage<BottomUpCheckpoint>),
+
+    // TODO
+    TopDown,
+}
+
+/// A message relayed by a user on the current subnet.
+///
+/// The relayer pays for the inclusion of the message in the ledger,
+/// but not necessarily for the execution of its contents.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct RelayedMessage<T> {
+    /// The relayed message.
+    pub message: T,
+    /// The address (public key) of the relayer in the current subnet.
+    pub relayer: Address,
+    /// The nonce of the relayer in the current subnet.
+    pub sequence: u64,
+}
+
+/// Relayed messages are signed by the relayer, so we can rightfully charge them message inclusion costs.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct SignedRelayedMessage<T> {
```

Comment:
How do we expect this `SignedRelayedMessage` to be triggered by users? I think we may have an issue here, as any bottom-up protocol we come up with would have to be compatible with Filecoin mainnet, i.e. Lotus. 

How I was imagining this to work, is that users directly send the `CertifiedMessage` to the parent inside an `eth_sendTransaction`, and the verification is performed on-chain in the subnet actor, which means that all the information required for the verification should be already included in the message. This means that the child subnets validators would have to tailor the `CertifiedMessage` in their subnet before it can be propagated up (maybe this is what you are thinking of doing and I am missing it from the type definitions). 

 For L2+ we can use this approach of relay+execution, but in L2->L1 bottom up interactions we may just be able to send a raw transaction against the ETH RPC to submit the checkpoint.

Reply:
> How do we expect this SignedRelayedMessage to be triggered by users? 

It would be sent by the IPC Actor _iff_ it knows the parent is a Fendermint node as well, because these messages don't work for Lotus at all. But with Lotus you also can't use the IPLD Resolver, so it's completely different. 

This is unfortunate, and I agree this whole thing with IPLD Resolution and CIDs is a bit forced in order to make use of IPLD in the first place. It's possible that if message size is a problem we should just have gone with a Merkle root in the checkpoint and Merkle proof in relayed messages, simples, at the cost of multiple relayers including the same transaction payload multiple times.

So, we have the well-known ID of the gateway actor where these messages are going, so in theory Fendermint could inspect all FVM messages going to that actor and try to deserialize them as the EVM calldata payload that should contain an `ethers` generated DTO. It would have to be a different one from the `BottomUpCheckpoint` in `Checkpoint.sol` you pointed at, because that one already has fully messages inside; if we want to use CIDs, that is. So it could be the equivalent of a `CertifiedMessage`, ready to be validated, with an opaque CID field to be resolved by the time of execution, at which point it can be turned into the Solidity `BottomUpCheckpoint`. 

That way we can drop this `SignedRelayedMessage`. It's existence is to make it clear that we are separating the act of relaying from the act of execution, which forces us to deal with the problem of who should pay for which part - the relayer pays for the validation, the inner message is paid for by its own sender(s). It should be possible to turn it around and lift the FVM message handling into Fendermint by parsing as well.

In any case, Lotus would have a different message format there which includes payloads.